src/libcore/fmt/num.rs

```diff
@@ -256,6 +257,161 @@ macro_rules! impl_Display {
     };
 }
 
+macro_rules! impl_Exp {
+    ($($t:ident),* as $u:ident via $conv_fn:ident named $name:ident) => {
+        fn $name(
+            mut n: $u,
+            is_nonnegative: bool,
+            upper: bool,
+            f: &mut fmt::Formatter<'_>
+        ) -> fmt::Result {
+            let mut exponent = 0;
```

Comment:
Ok, I can see what you were going for in the original implementation splitting out the initial exponent. I think it would be good to put this first part of the algorithm that does the initial work of figuring out precision in a block:

```rust
let (mut n, mut exponent, trailing_zeros, added_precision) = {
    ..
};
```